npe2/implements.py

```diff
@@ -0,0 +1,196 @@
+        return super().generic_visit(node)
+
+    def visit_ImportFrom(self, node: ast.ImportFrom) -> Any:
+        for alias in node.names:
+            self._names[alias.asname or alias.name] = f"{node.module}.{alias.name}"
+        return super().generic_visit(node)
+
+    def visit_FunctionDef(self, node: ast.FunctionDef) -> Any:
+        self._find_decorators(node)
+
+    def visit_ClassDef(self, node: ast.ClassDef) -> Any:
+        self._find_decorators(node)
+
+    def _find_decorators(self, node: Union[ast.ClassDef, ast.FunctionDef]):
+        for call in node.decorator_list:
+            if isinstance(call, ast.Call) and isinstance(call.func, ast.Attribute):
+                val = call.func.value
+                _names = []
+                while isinstance(val, ast.Attribute):
+                    _names.append(val.attr)
+                    val = val.value
+                if isinstance(val, ast.Name):
+                    _names.append(val.id)
+                    call_name = ".".join(reversed(_names))
+                    if self._names.get(call_name) == __name__:
+                        self._store_contrib(call.func.attr, call.keywords, node.name)
+        return super().generic_visit(node)
+
+    def _store_contrib(self, type_: str, keywords: List[ast.keyword], name: str):
+        # this can also be taken from schemas
+        _map: Dict[str, Tuple[Type[BaseModel], str]] = {
+            "writer": (contributions.WriterContribution, "writers"),
+            "reader": (contributions.ReaderContribution, "readers"),
+            "sample_data_generator": (contributions.SampleDataGenerator, "sample_data"),
+            "widget": (contributions.WidgetContribution, "widgets"),
+        }
+        Cls, contrib_name = _map[type_]
+        contrib = Cls(**self._store_command(keywords, name))
+        existing: List[dict] = self._contrib_points.setdefault(contrib_name, [])
+        existing.append(contrib.dict(exclude_unset=True))
+
+    def _store_command(self, keywords: List[ast.keyword], name: str) -> Dict[str, Any]:
+        kwargs = {str(k.arg): ast.literal_eval(k.value) for k in keywords}
+        cmd_kwargs = {k: kwargs.pop(k) for k in list(kwargs) if k in COMMAND_PARAMS}
+        cmd_kwargs["python_name"] = self._qualified_pyname(name)
+        cmd = contributions.CommandContribution(**cmd_kwargs)
+        if cmd.id.startswith(self.plugin_name):
+            n = len(self.plugin_name)
+            cmd.id = cmd.id[n:]
+        cmd.id = f"{self.plugin_name}.{cmd.id.lstrip('.')}"
+        cmd_contribs: List[dict] = self._contrib_points.setdefault("commands", [])
+        cmd_contribs.append(cmd.dict(exclude_unset=True))
+        kwargs["command"] = cmd.id
+        return kwargs
+
+    def _qualified_pyname(self, obj_name: str) -> str:
+        return f"{self.module_name}:{obj_name}"
+
+
+def compile(
+    path: Union[ModuleType, PathLike], plugin_name: str, module_name: str = ""
+) -> dict:
```

Comment:
I'd love to see docstrings for all these, even the private ones. As I'm super unfamiliar with all the patterns being used here, the code is very hard to parse. Docstrings could make it easier for newcomers to follow/fix/improve.

Reply:
will do!  this was "done enough" for a general concept review, but i'll fill out the details (type stubs and docs) before anything merges